### ISSUE TITLE
A stream blocks other queries to Sqlite

### ISSUE THREAD

**Issue by emilis** [tgriesser/knex#870]
I was trying to use a `stream()` to SELECT some rows from an Sqlite database, `ratelimit()` it with Highland.js and process them one by one.

However all subsequent SELECTs to the database are paused: their promises neither throw errors, nor return anything. Also there is no debug output.

If this is due to how Sqlite works, there should be a warning somewhere (at least in documentation).

In other case this should be fixed.

Note: I chose a stream because I want to avoid having to load the whole result into memory (could be quite large) and also because I need to rate-limit its processing: I am querying an external API for the records in the DB that are missing data from the API.

**Comment by tgriesser** [comment]
Yeah this is basically a limitation of how the node-sqlite3 driver works, it doesn't handle backpressure on large selects so everything is fetched into memory, the stream only exists to keep a consistent api with other dbs.

https://github.com/mapbox/node-sqlite3/issues/375

I could put a warning in the docs for this

**Comment by emilis** [comment]
I am sorry, but now I am confused: if everything is fetched into memory for a large select, why does streaming it blocks the following selects?

**Comment by mercmobily** [comment]
@emilis  Not exactly "timely" I know, but maybe you would still be interested?

The issue is that our SQLite "streaming" isn't actually streaming - it's loading the entire result set into memory first. Looking at the `_stream` method in lib/dialects/sqlite3/index.js:157-176:

```javascript
_stream(connection, obj, stream) {
  return client
    ._query(connection, obj)
    .then((obj) => obj.response)
    .then((rows) => rows.forEach((row) => stream.write(row)))
    .catch(function (err) {
      stream.emit('error', err);
    })
    .then(function () {
      stream.end();
    });
}
```

The problem is clear: we call `_query()` which fetches the entire result set into memory, then iterate over it to write to the stream. This completely defeats the purpose of streaming.

1. SQLite uses a single connection pool (`{ min: 1, max: 1 }`) - see line 216 in the same file
2. Large queries hold that connection until the entire result set is loaded
3. Other queries have to wait indefinitely

The node-sqlite3 driver doesn't support true streaming with backpressure, but our current implementation makes it worse by not warning users about this limitation.

Here's how I'd fix it:

**Quick fix**: Add a warning to the docs right next to the existing PostgreSQL streaming note explaining that SQLite streaming loads the full result set into memory and may block other queries.

**Better fix**: Add a runtime warning when `.stream()` is called on SQLite, similar to our existing `useNullAsDefault` warning:
```javascript
this.logger.warn(
  'SQLite streaming loads the entire result set into memory and may block other queries. ' +
  'Consider using pagination with LIMIT/OFFSET for large datasets.'
);
```

**Ideal solution (only in theory)**: After researching other libraries, LIMIT/OFFSET chunking would actually make performance worse for large datasets. The proper solution would be cursor-based pagination using `WHERE id > lastSeenId` instead of OFFSET, but this would require a new API design rather than retrofitting the existing `.stream()` method.

**What other libraries do:**
- PostgreSQL/MySQL: True driver-level streaming (pg-query-stream)  
- Prisma: Cursor pagination at ORM level, not streaming
- TypeORM: No SQLite streaming - throws explicit error
- Sequelize: Manual chunking workarounds only

The OFFSET-based approach I initially suggested would perform worse than the current implementation for large datasets (studies show 10+ minutes for 3.5GB with high offsets). A real solution needs cursor-based chunking with a different API.

I think we should definitely do the documentation warning immediately since this has been misleading users for years. The runtime warning would be a good follow-up, and cursor-based streaming would need to be a separate feature with a different API design.
